README.md
<h1 align="center">
  <br>
  <a href="http://liquid.westrocketry.com/index.html"><img src="http://liquid.westrocketry.com/images/patch.png" alt="Madison West Rocketry" width="200"></a>
  <br>
  SLI 2018 Liquid Vibrations
  <br>
</h1>

<h4 align="center">The software running on the payload of Madison West Rocketry's 2018 <a href="https://www.nasa.gov/audience/forstudents/studentlaunch/home/index.html" target="_blank">SLI project</a>.</h4>

## Madison West Rocketry (MWR)
Madison West Rocketry is a group of students who are studying the dampening effects of liquid vibrations in rockets. This repo is for the coding and keyboardsmashing that is needed to get useful data from the project and have a great student launch this year.

## Project Description
Welcome to our repository.This is the software for our
rocket payload experiment titled "Study of Damping Efficiency in Various Liquids." Considerable
effort was made to keep only the nessesary files for this repository. For any test files and examples
please go to [SLI2017_Hurgus_Madison](https://github.com/94KeyboardsSmashed/SLI2017_Hurgus_Madison).

## Project Gershwin:
- (G)ood
- (E)mperical
- (R)okketi
- (S)hip
- (H)ull
- (W)ith view to measure
- l(I)quid
- (N)atural frequences

## Prerequisites
In order to install the dependancies for this program do (sh install.sh). If problems arise, manually install
the following dependancies

### Accelerometer Dependancies
- python-smbus
- python3-smbus
- RPi.GPIO
- cap1xxx

### Bluetooth Dependancies
- PyBluez

### Neopixel Depenancies
- build-essential 
- python-dev 
- git 
- scons
- swig
- The rpi\_ws281x repo (https://github.com/jgarff/rpi_ws281x.git)

## Data Collection
The files in the data collection are the ones that will be used to read accelerometer readout from the payload
sensors. Files include a revamped portion of the rpi_ws281x library to control neopixel rings, and a accelerometer
readout library that interprets the accelerometers that will be attached in an i2c interface. The files VDD.py 
and GND.py will be used to initiate the payload experiment and the readout will be put onto log.txt files. 
### To run the whole setup 
```
sudo sh run_experiment.sh
```
### To run individual files
```
sudo python3 vdd.py
```
or
```
sudo python3 gnd.py
```
### To run individual files and write to log
```
sudo python3 gnd.py > loggnd.txt
```
or
```
sudo python3 vdd.py > logvdd.txt
```
### More than two files can be run and logged like this:
```
sudo nice -20 python3 gnd.py > loggnd.txt & python3 vdd.py > logvdd.txt
```

## Sockets
These are the bluetooth socket programs that we will use to transfer data between the pi's. Currently under
construction.

## License
This software and all associated files are under the MIT license.

## To do
- Integrate bluetooth into data collection?
- Ensure that raspberry pis sync to online time when connected to internet (bash script?)
- Ensure that all raspberry pis start recording when one accelerometer detects launch (or 2 or 3?)
- Create an easy way to switch the pis from lanch mode to rocket fair mode
- Test hardware for heat
- Test hardware for stress

## Other MWR repos
- [MWR_Web](https://github.com/googites/MWR_Web-/blob/master/README.md)
- [SLI2017_Hurgus_Madison](https://github.com/94KeyboardsSmashed/SLI2017_Hurgus_Madison)
- [MWR_PlasmaThruster](https://github.com/googites/MWR_PlasmaThuster)

## Authors
- Benton Likos
- Hyun-seok Chang




CHANGELOG.md
# Changelog
This changelog will document all notable changes to this file as per NASA 
request to keep detailed documentation for our Flight Readiness Review (FRR).

## [Beta v0.2.5] - 2018-03-14
### Changed
- setting.py max data file recording length changed to 5 mb
### Added
- Added buzzer component after landing
- Added \x04 tag to terminate readout
- Created new directory Data_Logs to hold backup log files
- Added backup data logs that save to Data_Logs directory with linux timestamps
- Added additional backup data logs that save to ../../Backup_Data directory with uuid
- Added easily configureable variables in setting.py in case of recording file change.

## [Beta v0.2.2] - 2018-03-05
### Changed
- Made .txt readout at startup
### Added
- Added buzzer component code to gnd.py
- Added buzzer channel control code in setup.py

## [Beta v0.2.1] - 2018-03-04
### Changed
- Reduced second file startup file amount from 2000 bytes to 1000 bytes
- Increased takeoff threashold from 5 to 7 gs
- Made circular data buffer respond to changes in measurement units

## [Alpha Release v0.2.0] - 2018-03-02
### Added
- Added Gershwin backronym explainer in README
- Added Other MWR repos section
- Added authors seections
### Changed
- Made the readings default measure in gs
- Circular buffer now in acceptable to gforce input
- Made Readme cleaner and added pictures
### Other
- Hardware assembled for first prototype launch on 2018-03-03

## [Alpha Release Candidate v0.1.1] - 2018-02-05
### Added
- Added easy X, Y, Z calibration values in settings.py.
### Changed
- install.sh overhauled to install dependancies to python3.
- Updated readme to do list.

## [Alpha Release Candidate v0.1.0] - 2018-02-04
### Added
- Added second trigger to unwind circular buffer when other accelerometers are unwound.
### Removed
- Removed file startup.py
- Removed all references to neopixels in vdd.py and gnd.py.
- Removed neopixel constant values from settings.py.
### Changed
- Restructured changelog format
- run_experiment.sh now prints "Working..." instead of executing startup.py.
- Max memory limit that stops gnd.py based on the size of loggnd.txt shortened from 5 MB to 2 MB (around 6:40 min).
- Max memory limit that stops vdd.py based on the size of logvdd.txt shortened from 5 MB to 2 MB (around 6:40 min).
- Takeoff Threshold constant in settings.py increased from 3gs to 5gs.
- Renamed accelerometer with address (0x53) from INDUS to YANGTZE.
- Range of accelerometers increased from +-8g to +-16gs.

## [Alpha v0.0.4] - 2018-02-03
### Added
- libbluetooth-dev added in install.sh as dependency.
- Added code to bluez\_client to find and connect with pis running bluez_server [WIP].
- Added additional examples in readme.
- Added CHANGELOG.md to document changes.
### Changed
- Max memory limit that stops vdd.py based on the size of logvdd.txt enlarged from 1 MB to 5 MB.
- Max memory limit that stops gnd.py based on the size of loggnd.txt enlarged from 1 MB to 5 MB.
- Updated module intro strings to contain more accurate information about the wiring of sparkfun accelerometers.
- Updated neopixel wiring specifications on module intro strings.

## [Alpha v0.0.3] - 2018-02-01
### Added
- Added a max memory limit that stops vdd.py when logvdd.txt file exceeds 1 MB.
- Added a max memory limit that stops gnd.py when loggnd.txt file exceeds 1 MB.
- Added code snippets in readme file
### Changed
- LANDING_THRESHOLD contant in settings.py set to 0 for testing purposes.
- Brightness of neopixel now depends fully on brightness value set at class init.
- startup.py fully integrated into run_experiment.sh file. Now runs properly at startup 
- Neopixel init brightness lowered from 255 to 16
- Changed accelerometer hardware from Adafruit brand to Sparkfun brand
- Expanded To Do section of readme file

## [Alpha v0.0.2] - 2018-01-31
### Added
- Proper markdown tags for readme.
- Added settings.py file - central file to change all variables.
### Removed
- Removed the vdd.py and gnd.py's need for neopixel connections.
### Changed
- Docstrings are now updated to the numpy docstring format.
- All files with global variables now have variables in the settings.py file.
- Module intro comments expanded and explains more.
- startup.py now runs only one neopixel ring with test times seperated by a magenta color wipe.

## [Alpha v0.0.1] - 2018-01-30
### Added 
- Licensed code under MIT license.
- Added Readme file with improper title tags.
- Added file install.sh - execute to install all dependancies.
- Added file bluez_client.py - basic bluetooth connection functionality for client.
- Added file bluez_server.py - basic bluetooth connection functionality for server host.
- Added file gnd.py - to read accelerometer data on (0x53) i2c address.
- Added file vdd.py - to read accelerometer data on (0x1D) i2c address.
- Added file run_experiment.sh - to execute both gnd.py and vdd.py and logs data in .txt files.
- Added file startup.py - experimental initiation file to execute on startup. Not yet implimented.
- Added file neopixel_lib.py - the neopixel library derived from previous github page.
- Added file raspi\_accel_lib.py - the accelerometer library derived from previous github page.
- Added file install.sh - executable file that will download all libraries.


Data_Collection/gnd.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#Python Data Collection Module

"""
Created on Tue May 9 12:13:27 2017

@author: Hyun-seok

IMPORTANT: REQUIRES ROOT ACCESS TO RUN.

Reads and interprets data from an ADXL 345 triaxial accelerometer and uses
a Adafruit 24 pixel neopixel ring to determine payload status.

Examples:
Run with the following lines on the command prompt:
    $ sudo python gnd.py

Run with this command to write to a .txt file:
    $ sudo nice -20 python gnd.py > logGND.txt

Wiring Details:

    Connect neopixel to ground, 5v, and physical pin 12 (gpio pin 18) (0)
    Connect neopixel to ground, 5v, and physical pin 33 (gpio pin 13) (1)

    GPIO 1 setup: ADXL - Pi (0x53)

    GND, SDO - GND
    VCC, CS - 3V3
    SDA - SDA (GPIO 2)
    SCL - SCL (GPIO 3)

    GPIO 2 setup: ADXL - Pi (0x1D)

    GND - GND
    VCC, CS, SDO - 3V3
    SDA - SDA
    SCL - SCL
"""

import sys
import collections
import time
from os import path
import raspi_accel_lib
import settings as st
import RPi.GPIO as GPIO

if __name__ == '__main__':
    # init global variables
    CIRCULAR_BUFF = collections.deque(maxlen=st.MAXLEN)
    AVG_BUFF = collections.deque(maxlen=st.ACCEL_RESPONSE)
    RESTING = 0

    # Define accelerometers (named after rivers)
    YANGTZE = raspi_accel_lib.ADXL345(st.CAL_X, st.CAL_Y, st.CAL_Z, 0x53)

    # Startup Accelerometer
    YANGTZE.accel_startup(st.GFORCE)

    # Initalize Buzzer:
    GPIO.setmode(GPIO.BCM)
    GPIO.setwarnings(False)
    GPIO.setup(st.CHANNEL, GPIO.OUT)

    # Initalize .txt file by writing headers
    print('#Time,X,Y,Z')
    print('#{}'.format(YANGTZE.string_output(st.GFORCE)))
    sys.stdout.flush()

    # Store up data in circular buffer on launch pad and
    # flush when launched.
    start = time.time()
    counter = 1
    state = False

    while True:
        if time.time() - start > 0.1:
            start = time.time()
            counter = counter - 0.1
        
        if counter <= 0 and state == False:
            counter = 0.1
            state = True
        
        if counter <=0 and state == True:
            counter = 1
            state = False

        GPIO.output(st.CHANNEL, state)

        CIRCULAR_BUFF.append(YANGTZE.string_output(st.GFORCE))
        # If this accelerometer or other accelerometers in network detect launch. Very rudimentary at the moment.
        if YANGTZE.accel_magnitude(True) > st.TAKEOFF_THRESHOLD or path.getsize('logvdd.txt') > 1000:
            BUFFER_DATA = list(CIRCULAR_BUFF)
            print('\n'.join(BUFFER_DATA))
            sys.stdout.flush()
            break

    # Record Data until vehicle is deemed to be "landed"
    while True:
        print(YANGTZE.string_output(st.GFORCE))
        sys.stdout.flush()

        if YANGTZE.accel_magnitude(True) < st.LANDING_THRESHOLD:
            RESTING += 1
        elif st.LANDING_SENSE < 0:
            RESTING = 0
        else:
            RESTING -= st.LANDING_SENSE

        if RESTING <= 0:
            RESTING = 0

        if RESTING >= st.RESTING_THRESHOLD:
            print("#Landed")
            print("\x04")
            sys.stdout.flush()
            break
			
        if path.getsize('loggnd.txt') > st.MEM_MAX:
            print('# Memory Stop')
            print("\x04")
            sys.stdout.flush()
            break
        
    while True:
        if time.time() - start > 0.1:
            start = time.time()
            counter = counter - 0.1
        
        if counter <= 0 and state == False:
            counter = 2
            state = True
        
        if counter <=0 and state == True:
            counter = 5
            state = False

        GPIO.output(st.CHANNEL, state)
